File: go/embeddings/text_embedder.go
// Package embeddings provides interfaces and implementations for text and visual embedding.
package embeddings

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

// TextEmbedder generates text embeddings via an HTTP embedding API.
type TextEmbedder struct {
	BaseURL   string
	APIKey    string
	Model     string
	Dimension int
	client    *http.Client
}

// NewTextEmbedder creates a TextEmbedder targeting the given API.
func NewTextEmbedder(baseURL, apiKey, model string, dimension int) *TextEmbedder {
	return &TextEmbedder{
		BaseURL:   baseURL,
		APIKey:    apiKey,
		Model:     model,
		Dimension: dimension,
		client:    &http.Client{},
	}
}

// Embed embeds a single text string. Returns a slice of float64.
func (t *TextEmbedder) Embed(ctx context.Context, text string) ([]float64, error) {
	results, err := t.EmbedBatch(ctx, []string{text})
	if err != nil {
		return nil, err
	}
	if len(results) == 0 {
		return nil, fmt.Errorf("empty embedding result")
	}
	return results[0], nil
}

// EmbedBatch embeds multiple texts.
func (t *TextEmbedder) EmbedBatch(ctx context.Context, texts []string) ([][]float64, error) {
	body := map[string]interface{}{
		"input": texts,
		"model": t.Model,
	}
	data, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, "POST", t.BaseURL+"/embeddings", bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if t.APIKey != "" {
		req.Header.Set("Authorization", "Bearer "+t.APIKey)
	}

	resp, err := t.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("embedding request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("embedding API error %d: %s", resp.StatusCode, string(respBody))
	}

	var result struct {
		Data []struct {
			Embedding []float64 `json:"embedding"`
		} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decoding embedding response: %w", err)
	}

	vectors := make([][]float64, len(result.Data))
	for i, d := range result.Data {
		vectors[i] = d.Embedding
	}
	return vectors, nil
}

// GetDimension returns the embedding dimension.
func (t *TextEmbedder) GetDimension() int {
	return t.Dimension
}

File: rust/src/embeddings/text_embedder.rs
//! Text embedding via HTTP API.
//!
//! Calls an OpenAI-compatible embedding API endpoint to produce vector
//! representations of text. Used for semantic similarity search in the
//! vector store.

use async_trait::async_trait;
use log::info;
use reqwest::Client;
use serde::{Deserialize, Serialize};

use crate::config::MemoryConfig;

/// Trait for text embedding backends.
#[async_trait]
pub trait TextEmbedderTrait: Send + Sync {
    /// Embed a single text string.
    async fn embed(&self, text: &str) -> Result<Vec<f64>, Box<dyn std::error::Error + Send + Sync>>;

    /// Embed multiple texts.
    async fn embed_batch(&self, texts: &[String]) -> Result<Vec<Vec<f64>>, Box<dyn std::error::Error + Send + Sync>>;

    /// Return the dimension of the embeddings.
    fn dimension(&self) -> usize;
}

#[derive(Debug, Serialize)]
struct EmbeddingRequest {
    model: String,
    input: Vec<String>,
}

#[derive(Debug, Deserialize)]
struct EmbeddingResponse {
    data: Vec<EmbeddingData>,
}

#[derive(Debug, Deserialize)]
struct EmbeddingData {
    embedding: Vec<f64>,
}

/// HTTP-based text embedder using an OpenAI-compatible embedding API.
pub struct TextEmbedder {
    model_name: String,
    api_url: String,
    api_key: Option<String>,
    client: Client,
    dim: usize,
}

impl TextEmbedder {
    pub fn new(config: &MemoryConfig, api_url: &str, api_key: Option<String>, dim: usize) -> Self {
        info!("Initializing text embedder with model: {}", config.text_embedding_model);
        Self {
            model_name: config.text_embedding_model.clone(),
            api_url: api_url.to_string(),
            api_key,
            client: Client::new(),
            dim,
        }
    }

    pub fn with_model(model_name: &str, api_url: &str, api_key: Option<String>, dim: usize) -> Self {
        Self {
            model_name: model_name.to_string(),
            api_url: api_url.to_string(),
            api_key,
            client: Client::new(),
            dim,
        }
    }

    async fn call_api(&self, input: Vec<String>) -> Result<Vec<Vec<f64>>, Box<dyn std::error::Error + Send + Sync>> {
        let body = EmbeddingRequest {
            model: self.model_name.clone(),
            input,
        };

        let mut req = self.client.post(&self.api_url).json(&body);
        if let Some(key) = &self.api_key {
            req = req.bearer_auth(key);
        }

        let resp: EmbeddingResponse = req.send().await?.json().await?;
        Ok(resp.data.into_iter().map(|d| d.embedding).collect())
    }
}

#[async_trait]
impl TextEmbedderTrait for TextEmbedder {
    async fn embed(&self, text: &str) -> Result<Vec<f64>, Box<dyn std::error::Error + Send + Sync>> {
        let results = self.call_api(vec![text.to_string()]).await?;
        results
            .into_iter()
            .next()
            .ok_or_else(|| "Empty embedding response".into())
    }

    async fn embed_batch(&self, texts: &[String]) -> Result<Vec<Vec<f64>>, Box<dyn std::error::Error + Send + Sync>> {
        self.call_api(texts.to_vec()).await
    }

    fn dimension(&self) -> usize {
        self.dim
    }
}
